### PR TITLE
Make sendTransactions sign and complete transaction requests

### LINKED ISSUE #1027: Get rid of transactions queue
# Problem

Because we don't have "sign later" functionality anymore, there is no reason to have transactions queue on go side.

# Implementation
-remove transactions queue functionality
-instead two `approve` methods implement one `sendTransaction`

image

### PR DESCRIPTION
**Pull request by sebastiandelgado**
The goal is to stop using the sign.PendingSignRequests queue in the Transactor and instead perform the signing and propagation of the transaction by istelf.

This was done by:
 - Getting the password for the current active account in backend.go and verifying it before passing it to the transactor
 - Having the transactor itself do the signing and propagation

Closes #1027

### CODE REVIEW

File: transactions/transactor_test.go
Comment (sebastiandelgado) [inline transactions/transactor_test.go:250]: Here the test was checking that the `sendHash` received when creating the transaction request in the old version of `SendTransaction` matched the response of the result of the request. Now that `SendTransaction` returns the result itself, and skips the sendHash, what would be a more sensible thing to be checking for? Can we include the expected result in each testcase? How would we figure out what that result should be?

File: api/backend.go
Comment (sebastiandelgado) [inline api/backend.go]: What is this SendTransaction call for? Not sure what arguments I should use here.

Comment (mandrigin) [reply api/backend.go]: this one is from RPC, if Clojure intercepts everything you don't need this `RegisterHandler` it at all
/cc @flexsurfer

File: transactions/transactor.go
Comment (mandrigin) [inline transactions/transactor.go]: I think we need to get rid of this function at all and just get the current account here or pass it to this function

Comment (sebastiandelgado) [reply transactions/transactor.go]: What replaces backend.SendTransaction then? How do we call transactor.SendTransaction directly from the RPC interface then?

Comment (mandrigin) [reply transactions/transactor.go]: In backend.SendTransaction you can get current active account and pass it into this function.

File: api/backend.go
Comment (sebastiandelgado) [inline api/backend.go]: The password needs to be passed to the backend.SendTransaction method in the backend.RegisterHandlers method. Otherwise, none of the RPC methods will work. Is it ok to not require the password in this case and just call the transactor.SendTransaction directly with the current active account?

File: transactions/transactor.go
Comment (sebastiandelgado) [inline transactions/transactor.go]: Do I need the signArgs argument anymore or should we just leave gas and gasprice in the regular txargs?

**Comment by sebastiandelgado** [comment]
Still Missing:

Update e2e rpc tests to reflect these changes
Update e2e tests using backend.SendTransaction directly

**Comment by sebastiandelgado** [comment]
Updated e2e tests.
Done:
 - Remove tests concerned queued transactions since they no longer exist
 - Re-register RPC handler for sendTransaction during account selection so that it uses the provided password when signing transactions
 - Update all e2e tests to use new sendTransaction method

File: api/backend.go
Comment (sebastiandelgado) [inline api/backend.go]: The current solution to passing the pwd to the sendtransaction method RPC handler is to add the password as an argument to the b.registerHandlers() method and calling it again when an account is selected.

File: t/e2e/transactions/transactions_test.go
Comment (sebastiandelgado) [inline t/e2e/transactions/transactions_test.go]: Before we were comparing the Tx hash returned from SendTransaction to the actual result. Now that the actual result is the only thing returned, what should we compare to that result? Right now I'm just checking for no error. IN the RPC tests I'm not checking anything else. In the tests that call sendTransaction directly, I'm only checking that the response is not null.

File: api/backend.go
Comment (mandrigin) [inline api/backend.go]: you don't need this handler whatsoever
we don't need to expose `eth_sendTransaction` anymore, it is Clojure's responsibility to call the right native method if `eth_sendTransaction` is called.

File: transactions/transactor.go
Comment (mandrigin) [reply transactions/transactor.go]: no idea, that wasn't me who introduced them :(

File: transactions/transactor_test.go
Comment (mandrigin) [reply transactions/transactor_test.go:250]: we don't know which hash the function will return in advance, so regular "this is not zeroes" check should suffice

**Review by mandrigin — CHANGES_REQUESTED** [review]
a few questions still

File: api/backend.go
Comment (sebastiandelgado) [reply api/backend.go]: Ok. so I'll remove the params.SendTransactionMethodName handler from here and remove the RPC tests for e2e transactor tests

**Comment by sebastiandelgado** [comment]
Addressed @mandrigin comments and removed WIP tag. This PR should be ready now. Will be working on getting CI to pass in case we have any trouble with that.

File: t/e2e/jail/jail_rpc_test.go
Comment (sebastiandelgado) [inline t/e2e/jail/jail_rpc_test.go:120]: Not sure if I should be removing these TestContractDeploymentRPC tests. Presumably contract creation should be called very differently via RPC after the changes in interface. Looking for input on how to do this.

**Comment by flexsurfer** [comment]
hey @sebastiandelgado could you point in code where you:

1) removed `"eth_sendTransaction"` and `"personal_sign"` web3 rpc methods
2) removed `"sign-request.queued"`  and `"sign-request.failed"`  signals
3) removed `approveSignRequest` and `approveSignRequestWithArgs` rpc methods
4) added `sendTransaction(transactionObj,password, callback)` and `signMessage(massageObj,password, callback)` rpc methods

thanks

**Comment by sebastiandelgado** [comment]
@flexsurfer 

1. removed `"eth_sendTransaction"` and `"personal_sign"` web3 rpc methods
- https://github.com/status-im/status-go/pull/1070/files#diff-900667c90030513e5563ef101e7a77daL301 but you'll see how I'm only getting rid of the handler for `"eth_sendTransaction"` and am leaving the one for `"personal_sign"` there. Did we not want to keep the current implementation for personal_sign?
 

2.  removed `"sign-request.queued"` and `"sign-request.failed"` signals
 - The signals are not called anymore because we are not calling `PendingRequests.Add` (https://github.com/status-im/status-go/blob/develop/sign/pending_requests.go#L41) or `Pending
Requests.Approve` (https://github.com/status-im/status-go/blob/develop/sign/pending_requests.go#L165). Let me know if you want me to add similar signals at some part of the process.

3. removed `approveSignRequest` and `approveSignRequestWithArgs` rpc methods
 - I don't think I did this. Are they missing from the build?

4. added `sendTransaction(transactionObj,password, callback)` and `signMessage(massageObj,password, callback)` rpc methods
 - This I definitely haven't done. Will get on it now

**Comment by sebastiandelgado** [comment]
@flexsurfer 

1. removed `"eth_sendTransaction"` and `"personal_sign"` web3 rpc methods
- https://github.com/status-im/status-go/pull/1070/files#diff-900667c90030513e5563ef101e7a77daL301 but you'll see how I'm only getting rid of the handler for `"eth_sendTransaction"` and am leaving the one for `"personal_sign"` there. Did we not want to keep the current implementation for personal_sign?
 

2.  removed `"sign-request.queued"` and `"sign-request.failed"` signals
 - The signals are not called anymore because we are not calling `PendingRequests.Add` (https://github.com/status-im/status-go/blob/develop/sign/pending_requests.go#L41) or `Pending
Requests.Approve` (https://github.com/status-im/status-go/blob/develop/sign/pending_requests.go#L165). Let me know if you want me to add similar signals at some part of the process.

3. removed `approveSignRequest` and `approveSignRequestWithArgs` rpc methods
 - I don't think I did this. Are they missing from the build?

4. added `sendTransaction(transactionObj,password, callback)` and `signMessage(massageObj,password, callback)` rpc methods
 - Why is `signMessage` required as part of these changes?  My understanding is that `sendTransaction` should be made as a native call to the backend. Is that not right?

File: t/e2e/jail/jail_rpc_test.go
Comment (sebastiandelgado) [reply t/e2e/jail/jail_rpc_test.go:120]: @mandrigin @flexsurfer I'm still confused as to whether this test should be removed. The test runs `web3.eth.contract.new()` in order to create the contract. This in turn calls the sendTransaction handler that I removed from backend.registerHandlers. 

What is the correct way to re-write this test? Whats the correct way to run a cell without calling backend.SendTransaction?

Other than that, this PR should be ready.

File: api/backend.go
Comment (divan) [inline api/backend.go]: In `sign` package, there is a (private) function `newErrResult`. Probably worth making it public and use here.

File: api/backend.go
Comment (divan) [inline api/backend.go:301]: Why is this handler removed?

This will effectively reroute `eth_sendTransaction` requests from dApps to the default  implementation of `SendTransaction`  in go-ethereum, right? Then `(*StatusBackend) SendTransaction()` method becomes obsolete. I'm a bit confused here.

cc @mandrigin

Comment (sebastiandelgado) [reply api/backend.go:301]: I was confused by this as well. I thought there were other calls to `backend.SendTransaction()` outside of the RPC one. In either case, if we were to keep this we'd need to pull re-register the handler using the password whenever an account is selected (so that it can call `backend.SendTransaction()`)

If we don't have this handler, where would status-react call SendTransaction from? @flexsurfer

Comment (divan) [reply api/backend.go:301]: Okay, so this handler should be removed, along with PersonalSign handler.

**Comment by divan** [comment]
Comments from @flexsurfer:

1) status go shouldn't listen or handle `"eth_sendTransaction"` and `"personal_sign"` these methods will be intercepted on status-react side
2) we need two new methods in status go interface `sendTransaction` and `signMessage` , instead of `approveSignRequest and approveSignRequestWithArgs` they should be removed
3) `"sign-request.queued"` and `"sign-request.failed"` signals not used anymore (edited)

**Review by adriacidre — APPROVED** [review]
Awesome work! LGTM when @divan 's comment on `eth_sendTransaction` gets clarified :)

**Comment by sebastiandelgado** [comment]
Thanks all for the comments.

Done:

- Added `SignMessage` and `SendTransaction` to status go interface
- Remove `approveSignRequest` and `approveSignRequestWithArgs` from status go interface
- Both `SignMessage` and `SendTransaction` now call statusBackend which in turn calls the corresponding service
- Created issue (#1097) to remove the PendingSignRequests queue once this PR is merged since its not being used anywhere else now

Missing:
 - Update tests for backend.go, services/personal/api.go and library.go.
 - Address a couple of issues I have commented on separately.

File: api/backend.go
Comment (sebastiandelgado) [inline api/backend.go]: @adambabik what is the right context to pass onto sign here? Couldn't find any calls to Sign before this PR so I don't have a reference of which one needs to be used.

Sign uses the context to pass along to api.rpcClient.CallContextIgnoringLocalHandlers

File: api/backend.go
Comment (sebastiandelgado) [inline api/backend.go:314]: @divan @andrey Do we also need to do the same for personalRecover? If we do, maybe we should split this PR up, as its getting pretty massive and taking a long time. We could do separate PRs for `sendTransaction`, `signMessage` and `ecRecover`

File: lib/library.go
Comment (sebastiandelgado) [inline lib/library.go:217]: @andrey I made similar interfaces for SendTransaction and SignMessage here. Basically unmarshalling JSON into the data structures that we use. Let me know if these will be workable on the status-react side or if we should modify them to be more like https://github.com/status-im/status-go/blob/develop/transactions/types.go#L77

File: api/backend.go
Comment (adambabik) [reply api/backend.go]: This is a little bit messy because `transactions.Transactor` has a field with a timeout. We can add one to `personal.PublicAPI` and remove passing context to `PublicAPI.Sign()` or keep it but wrap it in `PublicAPI.Sign()` implementation:
```go
func (api *PublicAPI) Sign(ctx context.Context, rpcParams Metadata) sign.Result {
  response := sign.EmptyResponse
  ctx, cancel := context.WithTimeout(ctx, api.rpcCallTimeout)
  defer cancel()
  err := api.rpcClient.CallContextIgnoringLocalHandlers(
  ctx,
  &response,
  params.PersonalSignMethodName,
  rpcParams.Data, rpcParams.Address, rpcParams.Password)
  // ...
}
```

File: api/backend.go
Comment (adambabik) [reply api/backend.go:314]: From status-react perspective, it would be better to have all these calls changed at once I believe. On status-go side, we can have many PRs merged to some feature branch and then the feature branch would be merged to `develop` without detailed review.

File: services/personal/api.go
Comment (adambabik) [inline services/personal/api.go:116]: Shouldn't this check still be around?

File: api/backend.go
Comment (adambabik) [reply api/backend.go:301]: I gave it some thought and I think it should stay. Even if `eth_sendTransaction` should be intercepted on status-react side, what will happen if actually, DApp manages to call `eth_sendTransaction` or will be called via HTTP when playing with `statusd`? I guess we should create a handled that returns an error if `eth_sendTransaction` is called.

File: api/backend.go
Comment (adambabik) [inline api/backend.go:315]: Similarly here.

File: lib/library_test_utils.go
Comment (adambabik) [inline lib/library_test_utils.go:810]: I would also create a unit test that checks failure scenario.

File: t/e2e/jail/jail_rpc_test.go
Comment (adambabik) [reply t/e2e/jail/jail_rpc_test.go:120]: The current approach should not allow calling `eth_sendTransaction` at all. If such an RPC call is received, we should return an error as it should be handled on status-react side by calling `SendTransaction()` function from `library.go`. I think this should be tested here. Similarly with `personal_*` APIs.

`backend.SendTransaction` uses some struct to send transactions or create contracts and only that struct should be unit tested. `backend.SendTransaction` can also have some unit/integration tests but I have seen mostly it covered.

File: t/e2e/transactions/transactions_test.go
Comment (adambabik) [inline t/e2e/transactions/transactions_test.go]: As mentioned in another comment, this test can stay but the error should be returned when calling RPC with `eth_sendTransaction` method.

File: t/e2e/transactions/transactions_test.go
Comment (adambabik) [inline t/e2e/transactions/transactions_test.go]: Similarly here.

File: t/e2e/transactions/transactions_test.go
Comment (adambabik) [inline t/e2e/transactions/transactions_test.go]: This test is actually different. We can rewrite it to use `s.Backend.SendTransaction()` with only `from` defined in the params.

File: transactions/transactor_test.go
Comment (adambabik) [inline transactions/transactor_test.go:161]: I think most of these test cases still make sense. They check various params passed in the params and they should be checked.

**Review by adambabik — COMMENTED** [review]
1. It looks like `sign/pending_requests.go` can be removed. We can do that in a different PR.
2. We can create a new feature branch called `tx-queue-removal` and use it as a base for smaller PRs. It has been already noticed that this PR can grow significantly.

File: api/backend.go
Comment (sebastiandelgado) [reply api/backend.go:315]: I don't think I understood your last point about creating a handler that returns an error if `eth_sendTransaction` is called. If we're keeping these then they should perform the expected actions and if we want to return an error for unrecognized/unavailable methods then we should have a separate handler for that no?

If we want to have these methods available via HTTP while playing with statusd then we should have them call `StatusBackend.SendTransaction/SignMessage` from the handler using the RPCParams.

File: transactions/transactor_test.go
Comment (sebastiandelgado) [reply transactions/transactor_test.go:161]: These tests were removed because we removed `Transactor.mergeSignTxArgsOntoSendTxArgs` ([https://github.com/status-im/status-go/pull/1070/files/312a47167309f5ea6d6582b0b7f51d2f2b7bcb08#diff-57c0d14258fd3623b3dd8f0b36b65775L209](url)). Since it was removed there is no need for testing the gas overriding. I removed it because now that we have one call (`sendTransaction`) instead of two (`sendTransaction` and then `approve`) we can just have all the gas info set in `transactions.SendTxArgs`.

File: services/personal/api.go
Comment (sebastiandelgado) [reply services/personal/api.go:116]: Yes. Addressed.

File: lib/library_test_utils.go
Comment (sebastiandelgado) [reply lib/library_test_utils.go:810]: Addressed

File: t/e2e/transactions/transactions_test.go
Comment (sebastiandelgado) [reply t/e2e/transactions/transactions_test.go]: Addressed

File: t/e2e/transactions/transactions_test.go
Comment (sebastiandelgado) [reply t/e2e/transactions/transactions_test.go]: Addressed

File: t/e2e/transactions/transactions_test.go
Comment (sebastiandelgado) [reply t/e2e/transactions/transactions_test.go]: Addressed

**Comment by sebastiandelgado** [comment]
Just addressed @adamb comments. Still missing
 - Update tests for backend.go, services/personal/api.go and library.go.
 - Split PR into multiple PRs that get merged into a feature branch. One PR per method (signMessage, recover, sendTransaction) and then another PR for removing the pending requests queue
 - Figure out if we're leaving the RPC handlers for each of the methods or creating a generic error handler for unavailable/unrecorgnizable methods

File: api/backend.go
Comment (adambabik) [reply api/backend.go:315]: The problem is that we expect a password when sending a transaction. Default `eth_sendTransaction` method does not require the password in the params. I am not sure what's gonna happen without a handler but we definitely need to secure that.

Yeah, another option is to add one more param which would be the password. However, that also does not seem secure as the password should not be sent in HTTP requests.

That's why, in my opinion, it's easier to just write at the moment:
```go
rpcClient.RegisterHandler(params. SendTransactionMethodName, UnsupportedMethodHandler)
```

**Comment by sebastiandelgado** [comment]
Closed as it will be replaced by #1101 (removes queue from transactor), #1102 (removes queue from personalSign) and a third PR that removes the PendingSignRequests code altogether.

**Pull request closed without merging**
